File: internal/git/ignore.go
package git

import (
	"io"
	"os"
	"path/filepath"
	"strings"
)

func AddIgnoreEntry(gitDir string, path string, dry bool) error {
	// Get relative path
	ignoreEntry, err := GetRelativePath(gitDir, path)
	if err != nil {
		return err
	}
	ignoreEntry = "/" + ignoreEntry // Add leading slash

	// Open the gitignore file, creating one if it doesn't exist
	ignoreFile, err := os.OpenFile(filepath.Join(gitDir, ".gitignore"), os.O_RDWR|os.O_APPEND|os.O_CREATE, 0644)
	if err != nil {
		return err
	}
	defer ignoreFile.Close()

	if dry {
		return nil
	}

	// Check if the file is already in the gitignore file
	fileBytes, err := io.ReadAll(ignoreFile)
	if err != nil {
		return err
	}
	if strings.Contains(string(fileBytes), ignoreEntry) {
		return nil
	}

	// Add the file to the gitignore file
	_, err = ignoreFile.WriteString("\n\n# Devious entry\n" + ignoreEntry)
	if err != nil {
		return err
	}

	return err
}

func RemoveIgnoreEntry(gitDir string, path string, dry bool) error {
	// Get relative path
	ignoreEntry, err := GetRelativePath(gitDir, path)
	if err != nil {
		return err
	}
	ignoreEntry = "/" + ignoreEntry // Add leading slash

	// Open the gitignore file
	// if the gitignore file doesn't exist, there's nothing to do
	ignoreFilePath := filepath.Join(gitDir, ".gitignore")
	if _, err := os.Stat(ignoreFilePath); os.IsNotExist(err) {
		return nil
	}
	ignoreFile, err := os.OpenFile(ignoreFilePath, os.O_RDWR, 0644)
	if err != nil {
		return err
	}
	defer ignoreFile.Close()

	// Find starting line to remove
	ignoreBytes, err := io.ReadAll(ignoreFile)
	ignoreFile.Seek(0, 0)
	if err != nil {
		return err
	}
	ignoreContents := strings.Split(string(ignoreBytes), "\n")
	for i, line := range ignoreContents {
		if line == ignoreEntry {
			err := removeLines(ignoreFile, i-1, 3)
			if err != nil {
				return err
			}
		}
	}

	if dry {
		return nil
	}

	return nil
}

File: src/internal/git/ignore.rs
use std::{fs::{File, OpenOptions}, path::PathBuf};
use crate::internal::git::repo;
use std::io::prelude::*;

pub fn add_gitignore_entry(git_dir: &PathBuf, path: &PathBuf) -> Result<(), std::io::Error> {
    // get relative path
    let ignore_entry_temp = match repo::get_relative_path(git_dir, path) {
        Ok(entry) => entry,
        Err(e) => return Err(e)
    };
    // Add leading slash
    let path = ignore_entry_temp.display().to_string();
    let ignore_entry = format!("/{path}");

    // open the gitignore file, creating one if it doesn't exist
    let ignore_file = git_dir.join(".gitignore");
    if !ignore_file.exists() {
       File::create(&ignore_file)?;
    }
    let contents = std::fs::read_to_string(&ignore_file).unwrap();
    if !contents.contains(&ignore_entry) {
        let mut file = OpenOptions::new()
        .write(true)
        .append(true)
        .open(ignore_file)
        .unwrap();

        if let Err(e) = writeln!(file, "\n\n# Devious entry\n{ignore_entry}" ) {
            eprintln!("Couldn't write to file: {}", e);
        }

    } // add ignore entry
    Ok(())
}